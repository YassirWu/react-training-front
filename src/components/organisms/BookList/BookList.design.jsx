import React from 'react';
import t from 'prop-types';

import { List, ListBody, ListHeader, ListBodyLine } from 'components';

import { Link } from 'react-router-dom';
import { getBookDetailRouteUrl } from 'routes';

const BookLine = ({
  bookId,
  name,
  numberOfPages,
  released,
  povCharacters,
  characters,
  LinkDetailComponent,
}) => (
  <ListBodyLine id={bookId}>
    <td>
      <LinkDetailComponent id={bookId}>{name}</LinkDetailComponent>
    </td>
    <td>{numberOfPages}</td>
    <td>{released}</td>
    <td>{povCharacters}</td>
    <td>{characters}</td>
  </ListBodyLine>
);
BookLine.propTypes = {
  bookId: t.number.isRequired,
  name: t.string.isRequired,
  numberOfPages: t.number.isRequired,
  released: t.string.isRequired,
  povCharacters: t.number.isRequired,
  characters: t.number.isRequired,
  LinkDetailComponent: t.node,
};
BookLine.defaultProps = {
  LinkDetailComponent: ({ children }) => <span>{children}</span>,
};

const renderListLines = book => (
  <BookLine
    bookId={book.id}
    name={book.name}
    numberOfPages={book.numberOfPages}
    released={book.released}
    povCharacters={book.povCharacters.length}
    characters={book.characters.length}
    LinkDetailComponent={({ id, children }) => <Link to={getBookDetailRouteUrl(id)}>{children}</Link>}
  />
);

const BookList = ({ books }) => (
  <List>
    <ListHeader>
      <th>Book</th>
      <th>Nb of pages</th>
      <th>Release date</th>
      <th>Pov characters</th>
      <th>Nb of characters</th>
    </ListHeader>
    <ListBody>
      {books.map(renderListLines)}
    </ListBody>
  </List>
);

BookList.defaultProps = {
  books: [],
};

BookList.propTypes = {
  books: t.arrayOf(
    t.shape({
      id: t.number.isRequired,
      name: t.string.isRequired,
      numberOfPages: t.number.isRequired,
      released: t.string.isRequired,
      povCharacters: t.arrayOf(t.string.isRequired),
      characters: t.arrayOf(t.string.isRequired),
    }),
  ),
};

export default BookList;
