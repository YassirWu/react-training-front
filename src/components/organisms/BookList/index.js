export { default } from './BookList.design';
