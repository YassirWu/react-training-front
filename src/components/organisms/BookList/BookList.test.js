import React from 'react';
import { shallow } from 'enzyme';

import BookList from './BookList.design';

describe('<BookList />', () => {
  const books = [
    {
      isbn: '978-0553103540',
      mediaType: 'Hardcover',
      name: 'A Game of Thrones',
      povCharacters: [],
      characters: []
    },
    {
      isbn: '978-0553108033',
      mediaType: 'Hardback',
      name: 'A Clash of Kings',
      povCharacters: [],
      characters: []
    },
    {
      isbn: '978-0553106633',
      mediaType: 'Hardcover',
      name: 'A Storm of Swords',
      povCharacters: [],
      characters: []
    },
    {
      isbn: '978-0976401100',
      mediaType: 'GraphicNovel',
      name: 'The Hedge Knight',
      povCharacters: [],
      characters: []
    },
  ];
  
  test('should renders BookList component', () => {
    const wrapper = shallow(<BookList books={books} />);
    expect(wrapper).toMatchSnapshot();
  });
  test('should return list of books', () => {
    const wrapper = shallow(<BookList books={books} />);
    expect(wrapper.find('BookLine')).toHaveLength(4);
  });
});
