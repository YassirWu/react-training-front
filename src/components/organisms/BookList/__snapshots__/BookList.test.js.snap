// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookList /> should renders BookList component 1`] = `ShallowWrapper {}`;
